

# Allen Brain Cell Atlas (ABC) Mouse Datasets

- Allen Brain Cell Atlas (ABC) Mouse Single cell datasets
    - Yao et. al. (2023) :  Two large gene expression datasets and a Spatial Gene expression dataset
    - Zhang et. al. (2023) : Three Spatial Gene expresion datasets
- Current Focus: Yao et. al. (2023)  Spatial Transcriptomics Dataset
    - ~4.0 million cells.
    - 500 gene panel MERFESH dataset.
- Dataset source:
    - 59 coronal sections (10 micron thick) covering the whole anterior to posterior extent of the brain (~200 micron apart) from a single adult male mouse. 
- With MERFISH assay, each cell also has a set of x, y, z coordinates:
    - x and y generated by rotating each section so that it is upright with midline approximately in the middle of the frame.
    - z coordinate takes into account the physical spacing between each section

# Downloading data
- ABC data is stored in AWS data strore.
- In order to download the data, AWS CLI needs to be installed.
- __download_abc_data__ function is available in the abc_mouse package to download the necessary data.
- After setting the appropriate abc_mouse.ABC_BASE variable, **Uncomment** the last line in the cell below to download the data in the ABC_BASE directory.

# Structure of a single cell Gene Expression dataset

![image.png](img/)

Gene expression datasets can be seen as composed of three tables :

- Gene Expression Data
- Cell meta data
   - Experiment: e.g. disease, organism, sex, donor, etc.
   - Region: e.g. layer, region, etc.
   - Stats & Analysis: e.g. ncounts, cell type, etc.
- Gene meta data
   - Avg. expression level, Is highly variable, etc.

```
# Load Cell metadata and Gene metadata 
manifest, file_meta = abc_mouse.merfish_files_meta()
cell_meta = abc_mouse.cell_metadata(file_meta)
gene_meta = abc_mouse.gene_metadata(file_meta)
```

## Whole Brain Taxonomy Data
The Allen Brain Cell Atlas includes an integrated transcriptomic taxonomy that contains
 - 5,322 clusters that are organized in a hierarchical manner
     - Top level of 34 classes
     - Second-level of 338 subclasses
     - Third-level of 1,201 supertypes and
     - Fourth-level of 5,322 types/clusters.

Example of classes:

     '01 IT-ET Glut',
     '02 NP-CT-L6b Glut',
     '03 OB-CR Glut',
     '04 DG-IMN Glut',
     '05 OB-IMN GABA',
     '06 CTX-CGE GABA',
     '07 CTX-MGE GABA',
     '08 CNU-MGE GABA',
     '09 CNU-LGE GABA'

### Construction of Exteneded Cell Metadata

- Load the taxonomy data for all the cells
- Extend the cell meta data to include the class, subclass and types information.

```
cluster_details, cluster_colors = abc_mouse.taxonomy_cluster(manifest)
cell_meta_ext = cell_meta.join(cluster_details, on="cluster_alias")
cell_meta_ext = cell_meta_ext.join(cluster_colors, on="cluster_alias")
```

## Allen Mouse CCF co-ordinates and parcellation
- Dataset source:
    - 59 coronal sections (10 micron thick) covering the whole anterior to posterior extent of the brain (~200 micron apart) from a single adult male mouse. 
- As mentioned in a MERFISH dataset, each cell also has a set of x, y, z coordinates.
- Allen Mouse Brain Common Coordinate Framework (CCFv3, Wang et al, 2020) is a 3D reference space of an average brain at 10um voxel resolution.
- Each section from the whole brain MERFISH dataset is mapped Allen CCFv3.
- Parcellaton division, structure and substructure are derived from the co-ordinates mapped to Allen CCF.

```
# Loading CCF Meta data
merfish_ccf_view_dir = abc_mouse.view_dir(manifest, "MERFISH-C57BL6J-638850-CCF")
cell_ccf = abc_mouse.cell_ccf_meta(os.path.join(merfish_ccf_view_dir, abc_mouse.PARCEL_META_DATA))
```

## Region Specific Ratios
- Inhibitory types are : 'Vip', 'Pvalb', 'SSt', 'Lamp5', 'Sst-Chodl', 'GABA-Other' (Inhibitory cells that are not one of Vip, Pvalb, SSt, SST-Chodl)
- Exitatory types are :  "IT", "ET", "CT", "NP", "Glut-Other"(Excitatory cells that are not one of IT, ET, CT or NP)
- IT types are further classified as "IT-CTX","IT-ENT" and "IT-Other"
- Based on the parcelletion information of the Allen CCF to find the layer specific ratios using _region_ccf_cell_types_ function
- First, we get data frame with flags set indicating the cell types and cell sub-types from the meta data
- Next, we group by the layers in sub-region to obtain summary statistics
- Finally, we compute the ratios in each layer 
- The above three steps is computed for each layer and merge all the data
```
# Obtain Region Specific CCF
region_cell_ccf, region_ei_ccf = abc_mouse.region_ccf_cell_types(cell_ccf, ["MOp", "MOs", "AUDp", "AUDd", "VISp", "VISl"])
# Merge the data 
cortex_ei_df = pd.concat([y for x,y in region_ei_ccf.items()])
cortex_ei_df

```

## Ratio data
- Ratio data is made available as a data frame with the columns indicating the fractions
- Key columns are
    1. 'fraction wi. region' : Fraction of cells in this layer within this region
    2. 'inhibitory fraction' : Fraction of cells that re inhibitory
    3. 'Vip fraction' : Among the inhibitory cells, fraction of Vip cells
    4. 'Lamp5 fraction' : Among the inhibitory cells, fraction of Lamp5 cells
    5. 'Sst-Chodl fraction' : Among the inhibitory cells, fraction of Sst-Chodl cells
    6. 'GABA Other fraction' : Among the inhibitory cells, fraction of cells that is none of Vip, Lamp5, SSt, SSt-Chodl
    7. 'IT fraction': Among the excitatory cells, fraction of cells that are intratelencephalic projecting.
    8. 'ET fraction': Among the excitatory cells, fraction of cells that are extratelencephalic projecting.
    9. 'CT fraction': Among the excitatory cells, fraction of cells that are corticothalamic
    10. 'NP fraction': Among the excitatory cells, fraction of cells that are non-projecting
    11. 'Glut Other fraction': Among the excitatory cells, fraction of cells that are not one of IT, ET, CT or NP
    12. 'IT-CTX fraction', 'IT-ENT fraction', 'IT-Other fraction' : Fraction of cells among the IT cell types


# Filter cell meta data by region of interest
- Subset the data from the region of interest
- Here we will use parcellation_structure information to filter the dataset to include only "VISp" cells 
- After filtering, we observe the dataset has 61884 cells
- Note the data includes all the genes, but only has the subet

### For V1 Model
- Select only cells with "parcellation_structure" assigned as "VISp' 
- Selected 61884 cells out of ~4 million cells

```
#
# Select only primary visual cortex cells
# 
visp_cell_ccf = cell_ccf.loc[cell_ccf["parcellation_structure"] == "VISp"]
visp_cells = visp_cell_ccf["cell_label"].values
visp_cell_meta = cell_meta_ext.loc[cell_meta_ext.index.isin(visp_cells)]
visp_edata = exp_data[visp_cells, :].to_memory()
# Remove blank genes from meta data
visp_gene_meta = abc_mouse.filter_invalid_genes(visp_edata.var,
                                                abc_mouse.valid_genes_symbols(exp_data.var))

```

## An Unified Data frame indexed by cell ids

Construct a unified data frame where each row is a cell and contains the following information

- Gene expression for all 500 genes
- Cell Meta data including class, sub-class etc.
- CCF Meta data including parcellation structure
- Can potentially divide by sections, however we select all the necessary selection


```
# Construct a data frame with 
section = visp_cell_meta[np.repeat(True, len(visp_cell_meta))]  # All the sections
vis_exp_data = abc_mouse.create_expression_dataframe(visp_edata, visp_gene_meta, section)
# Merge with CCF data
pcl_fields = [
    "parcellation_division",
    "parcellation_structure",
    "parcellation_substructure",
    "x_ccf",
    "y_ccf",
    "z_ccf",
]
uni_vis_data = vis_exp_data.join(cell_ccf.set_index("cell_label").loc[vis_exp_data.index][pcl_fields])
uni_vis_data
```

## Inhibitatory Fraction within Each Layer

- A V1 model can used the fraction of inhibitatory neurons as a parameters

## Steps to derive inhibitatory fraction from Spatial Transcriptomics dataset 
- Each cell is annoted with a class label and a CCF parcellation substructre 
- Annotation give in a spreadsheat [Class annotation](https://allen-brain-cell-atlas.s3-us-west-2.amazonaws.com/metadata/WMB-taxonomy/20231215/cl.df_CCN202307220.xlsx)
- ![image.png](attachment:faa5ff0d-ca98-4d1b-99b7-d99c3ed1e40e.png)
- To find this ihibitatory ratio, we assume:
    - The class label with GABA as inhibitory neurons
    - The class label with Glut as excitatory neurons
    - Identify the ratio of inhibitatory with respect to the total neurons for each layer


```
pred = uni_vis_data['class'].str.endswith("Glut") | uni_vis_data['class'].str.endswith("GABA")
# Select only GABA and Glut cells
visp_ei_data = uni_vis_data.loc[pred]
visp_ei_data.loc[:,["E"]] = visp_ei_data['class'].str.endswith("Glut")
visp_ei_data.loc[:,["I"]] = visp_ei_data['class'].str.endswith("GABA")
# Group by sub structure and find counts
visp_ei_data = visp_ei_data[["parcellation_substructure", "E","I"]]
visp_ei_ratio = visp_ei_data.groupby("parcellation_substructure").sum()
visp_ei_ratio["T"] = visp_ei_ratio["E"] + visp_ei_ratio["I"]
visp_ei_ratio["inhibitory fraction"] = visp_ei_ratio["I"]/visp_ei_ratio["T"]
visp_ei_ratio
```

## Fraction within Excitatory/Inhibitatory cell types

- Models use fraction of cell types within each excitatory and inhibitatory set.
- We examine two cell types Pvalb and Vip ratios based on the Gene expression Data

## Steps
- Select only the inhibitatory cell types (class "GABA")
- Among these selected cells, compute the average gene expression of each gene within each layer
- Compare the ratio of the gene expression to the ratio of the fraction

```
pred = uni_vis_data['class'].str.endswith("GABA")
# Select only GABA cells
visp_ib_data = uni_vis_data.loc[pred]
# Select the genes of interset
select_genes = [
        "Vip",
        "Pvalb",
]
# Find the average gene expression Values for the genes
visp_ib = abc_mouse.aggregate_by_metadata(visp_ib_data, visp_gene_meta.gene_symbol, "parcellation_substructure")
visp_ib = visp_ib[select_genes]
# Compute the metrics
visp_ib["Vip/Pvalb"] = visp_ib["Vip"] / visp_ib["Pvalb"]
visp_ib
```
